rpm/xcache.spec

```diff
@@ -40,6 +40,31 @@ Obsoletes: stashcache-daemon < 1.0.0
 %postun
 %systemd_postun_with_restart xcache-reporter.service xcache-reporter.timer xrootd-renew-proxy.service xrootd-renew-proxy.timer
 
+########################################
+%package -n xcache-consistency-check
+Summary: Consistency check for root files
+
+Requires: xz
+Requires: python-pip
+Requires: xrootd-server
+
+%description -n xcache-consistency-check
+%{summary}
+
+%post -n xcache-consistency-check
```

Comment:
You _really really_ shouldn't be doing this. Post-install scripts should not install additional software, especially via pip (where it is unmanaged by RPM). Find or make RPMs for your Python dependencies and install them that way.

Reply:
If you need to make RPM packages for these Python modules, the [packaging guidelines](https://docs.fedoraproject.org/en-US/packaging-guidelines/Python/) have instructions on what you can do to make things easier.

Reply:
Either that or else if your tests need to install stuff via pip, you can put those install commands in a separate script that the user can manually run.  That "takes away the offense" of having the post-install scriptlet itself do the non-rpm installs.  (My 2c)

Reply:
Thanks @edquist  but those dependencies are required for the main script to work not just for testing  and most of them if not all do not have an available rpm, so I'll have to create them.

Reply:
@matyasselmeci, thanks for your comments. I used to have a python installer (setup.py) that would install these dependencies when called from the spec file by 'py_install' but I think that would be the same, wouldn't it?
So what you are suggesting is for me to create rpms for every dependency, does that mean that these rpms would live in our repos as well?

Reply:
> Thanks edquist but those dependencies are required for the main script to work not just for testing and most of them if not all do not have an available rpm, so I'll have to create them.

Makes sense, sorry that makes it a hassle.

By the way (and i'm not suggesting this is a good idea, but, a thought experiment anyway), if the stuff from pip is just needed for the rpm scriptlet to run (but not for the package otherwise), another possibility might be to pip-install into a temporary directory (instead of a system location) that gets removed when the scriptlet finishes.  That would avoid the package leaving behind stuff that does not get managed by rpm.  But making rpm packages for the pip stuff also works.

Reply:
So this is what I've got with pipdeptree:
`(tree) Diegos-MBP:~ ddavila$ pipdeptree 
lz4==2.2.1
  - future [required: Any, installed: 0.18.2]
pyliblzma==0.5.3
uproot==3.11.2
  - awkward [required: >=0.12.0,<1.0, installed: 0.12.20]
    - numpy [required: >=1.13.1, installed: 1.16.6]
  - cachetools [required: Any, installed: 3.1.1]
  - numpy [required: >=1.13.1, installed: 1.16.6]
  - uproot-methods [required: >=0.7.0, installed: 0.7.3]
    - awkward [required: >=0.11.0, installed: 0.12.20]
      - numpy [required: >=1.13.1, installed: 1.16.6]
    - numpy [required: >=1.13.1, installed: 1.16.6]
xxhash==1.4.3
`
numpy does have an available rpm

Reply:
the alignment of the dependency tree  didn't show up well

Reply:
for my bad luck, the latest rpm available for numpy is the version 1.7 and I need at least 1.13 so numpy is back on the list of rpms to build

Reply:
Diego and I discussed this over Slack; what we're going to try is pip installing packages in the %install section into a separate directory, and then setting PYTHONPATH when running the script so it finds the packages.  So something like
`pip2 install --root=%{buildroot} --prefix=/usr/lib/xcache-consistency-check PACKAGE` in %install,

and
`Environment=PYTHONPATH=/usr/lib/xache-consistency-check` in the .service file.

Reply:
Does this need to be in the RPM?  Could we simply package this as a docker container?  The rpm is a timer / service that calls the docker container to scan a directory?  spitballing...

Reply:
How would that work with the stashcache containers? Sidecar? Docker-in-docker?

Reply:
Sidecar i suppose.  I'm not sure how it would work with Kubernetes (I know how it would work with something low level like docker-compose).  Can multiple containers mount the same storeage?

Reply:
I'm pretty sure you can bind-mount paths multiple times so that part would work... but I'm not sure admins would be happy about having to install docker on their caches just to run a cronjob...

Reply:
I tried @matyasselmeci idea of installing the python dependencies in a separate directory but I get this error when I try to install the rpm[1]. Normally the library on question (libgfortran-ed201abd.so.3.0.0) gets installed by numpy at: /usr/lib64/python2.7/site-packages/numpy/.libs/libgfortran-ed201abd.so.3.0.0 and apparently when it gets installed in a different location the rpm installation process cannot see it and complains about it.
I tried defining PYTHONPATH to the custom location of the library, before running 'yum install' but doesn't make any change. Any ideas?

[1]
[root@45be2d5a1993 /]# yum install rpms/x86_64/xcache-consistency-check-1.3.0-1.el7.x86_64.rpm               
Loaded plugins: fastestmirror, ovl
Examining rpms/x86_64/xcache-consistency-check-1.3.0-1.el7.x86_64.rpm: xcache-consistency-check-1.3.0-1.el7.x86_64
Marking rpms/x86_64/xcache-consistency-check-1.3.0-1.el7.x86_64.rpm to be installed
Resolving Dependencies
--> Running transaction check
---> Package xcache-consistency-check.x86_64 0:1.3.0-1.el7 will be installed
--> Processing Dependency: libgfortran-ed201abd.so.3.0.0(GFORTRAN_1.0)(64bit) for package: xcache-consistency-check-1.3.0-1.el7.x86_64
Loading mirror speeds from cached hostfile
 * base: mirror.steadfastnet.com
 * epel: d2lzkl7pfhq30w.cloudfront.net
 * extras: linux-mirrors.fnal.gov
 * updates: linux-mirrors.fnal.gov
--> Finished Dependency Resolution
Error: Package: xcache-consistency-check-1.3.0-1.el7.x86_64 (/xcache-consistency-check-1.3.0-1.el7.x86_64)
           Requires: libgfortran-ed201abd.so.3.0.0(GFORTRAN_1.0)(64bit)

Reply:
Now the build fails due to: BuildArch: noarch, I think numpy is installing arch dependent files and thus I would need to set BuildArch: x86_64 (I've tested it and works), unfortunately it seems that you cannot use different BuildArch for the different subpackages (at least not in an simple way) so all the packages would be build as x86_64, is this ok or should we reconsider separating this tool from the xcache package?

Reply:
You don't need to add `BuildArch: x86_64`, just drop `BuildArch: noarch`. And that's fine, no need to separate.